projects/optic/src/utils/cloud-specs.ts

```diff
@@ -0,0 +1,78 @@
+import { IChange, ResultWithSourcemap } from '@useoptic/openapi-utilities';
+import { OpticBackendClient } from '../client';
+import { computeChecksum } from './checksum';
+import { uploadFileToS3 } from './s3';
+import { ParseResult } from './spec-loaders';
+
+export const EMPTY_SPEC_ID = 'EMPTY';
+
+export async function uploadSpec(
+  apiId: string,
+  opts: {
+    spec: ParseResult;
+    client: OpticBackendClient;
+    tags: string[];
+  }
+): Promise<string> {
+  const spec_checksum = computeChecksum(opts.spec.jsonLike);
+  const sourcemap_checksum = computeChecksum(opts.spec.sourcemap);
+  const result = await opts.client.prepareSpecUpload({
+    api_id: apiId,
+    spec_checksum,
+    sourcemap_checksum,
+  });
+  if ('upload_id' in result) {
+    await Promise.all([
+      uploadFileToS3(
+        result.spec_url,
+        Buffer.from(JSON.stringify(opts.spec.jsonLike))
```

Comment:
You'll need to stable stringify these too, or S3 will reject the upload as not matching the checksum.